service/src/main/python/jcarbon/tensorflow/callbacks.py

```diff
@@ -125,6 +97,98 @@ def on_epoch_end(self, epoch, logs=None):
         self.reports[epoch] = self.last_report.to_frame().assign(
             epoch=epoch).set_index('epoch', append=True)
 
+class JCarbonExperimentCallback(JCarbonChunkingCallback):
+    def __init__(
+            self,
+            addr='localhost:8980',
+            period_ms=DEFAULT_PERIOD_MS,
+            signals=DEFAULT_SIGNALS,
+            chunking_period_sec=DEFAULT_PERIOD_SECS):
+        super().__init__(addr, period_ms, signals)
+        self.timestamps = {}
+        self.batch_timestamps = None
+
+    def on_epoch_begin(self, epoch, logs=None):
+        super().on_epoch_begin(epoch, logs)
+        self.curr_batch_timestamps = None
+    
+    def on_train_batch_begin(self, batch, logs=None):
+        self.batch_start = time.time()
+        super().on_train_batch_begin(batch, logs)
+        
+        
+    def on_train_batch_end(self, batch, logs=None):
+        curr = time.time()
+        super().on_train_batch_end(batch, logs)
+        self.batch_end = int((10**9 * curr))
+        self.batch_start = int((10**9 * self.batch_start))
+        if self.curr_batch_timestamps is None:
+            self.curr_batch_timestamps = [{'batch': batch, 'start': self.batch_start, 'end': self.batch_end}]
+        else:
+            self.curr_batch_timestamps.append({'batch': batch, 'start': self.batch_start, 'end': self.batch_end})
+
+    def on_epoch_end(self, epoch, logs=None):
+        super().on_epoch_end(epoch, logs)
+        if self.batch_timestamps is None:
+            self.batch_timestamps = pd.DataFrame.from_dict(self.curr_batch_timestamps)
+        else:
+            self.batch_timestamps = pd.concat([pd.DataFrame.from_dict(self.curr_batch_timestamps)])
+        self.timestamps[epoch] = self.batch_timestamps.assign(
+            epoch=epoch)
```

Comment:
I am not fully sure what is happening here but it looks incorrect (or has an invisible error). You are mutating the batch timestamps for some reason, which just looks plain wrong since all you need to do add the dataframe to the timestamps dict. Instead it looks like you created a lot of additional steps that were ultimately just bookkeeping.

Using the pattern I described above, you should just pass the list of dicts for the batch timestamps to `pd.DataFrame`. Note that `pd.DataFrame.from_dict` is for nested dictionaries, not lists of dicts.